File: Compass/Logger.cs
namespace Compass;

public static class Logger
{
    public enum LogLevel
    {
        Verbose = 0,
        Info = 1,
        Warning = 2,
        Error = 3
    }

    public static List<TextWriter> Writers { get; set; } = [Console.Out];

    public static LogLevel CurrentLogLevel = LogLevel.Verbose;

    public static void Log(string message, object obj, LogLevel level = LogLevel.Info, bool includeTimestamp = true)
    {
        Log($"{message}\n{obj.ToJsonString()}", level, includeTimestamp);
    }

    public static void Log(string message, LogLevel level = LogLevel.Info, bool includeTimestamp = true)
    {
        if (level < CurrentLogLevel) return;

        message = System.Text.RegularExpressions.Regex.Unescape(message);
        if (includeTimestamp)
        {
            message = $"[{DateTime.Now:HH:mm:ss}] {message}";
        }
        var text = message;

        LogToWriter(text);
    }

    public static void LogToWriter(string text)
    {
        if (Writers is not null)
        {
            foreach (var writer in Writers)
            {
                writer.WriteLine(text);
            }
        }
    }
}

File: src/logger.ts
export enum LogLevel {
  Verbose = 0,
  Info = 1,
  Warning = 2,
  Error = 3,
}

export class Logger {
  private static writers: Array<(text: string) => void> = [(text) => console.log(text)];
  private static _currentLogLevel: LogLevel = LogLevel.Verbose;

  static get currentLogLevel(): LogLevel {
    return this._currentLogLevel;
  }

  static set currentLogLevel(level: LogLevel) {
    this._currentLogLevel = level;
  }

  static setWriters(writers: Array<(text: string) => void>): void {
    this.writers = writers;
  }

  static log(message: string, level: LogLevel = LogLevel.Info, includeTimestamp: boolean = true): void;
  static log(message: string, obj: unknown, level: LogLevel, includeTimestamp: boolean): void;
  static log(
    message: string,
    objOrLevel?: unknown | LogLevel,
    levelOrIncludeTimestamp?: LogLevel | boolean,
    includeTimestamp: boolean = true
  ): void {
    let level: LogLevel;
    let finalMessage: string;

    if (typeof objOrLevel === "number" && objOrLevel in LogLevel) {
      // log(message, level, includeTimestamp)
      level = objOrLevel as LogLevel;
      finalMessage = message;
      if (typeof levelOrIncludeTimestamp === "boolean") {
        includeTimestamp = levelOrIncludeTimestamp;
      }
    } else if (objOrLevel !== undefined && typeof objOrLevel !== "number") {
      // log(message, obj, level, includeTimestamp)
      level = (levelOrIncludeTimestamp as LogLevel) ?? LogLevel.Info;
      finalMessage = `${message}\n${JSON.stringify(objOrLevel, null, 2)}`;
    } else {
      // Default case
      level = LogLevel.Info;
      finalMessage = message;
    }

    if (level < this._currentLogLevel) return;

    // Unescape the message (similar to C# Regex.Unescape)
    finalMessage = finalMessage
      .replace(/\\n/g, "\n")
      .replace(/\\r/g, "\r")
      .replace(/\\t/g, "\t")
      .replace(/\\\\/g, "\\");

    if (includeTimestamp) {
      const now = new Date();
      const timestamp = now.toTimeString().split(" ")[0]; // HH:mm:ss format
      finalMessage = `[${timestamp}] ${finalMessage}`;
    }

    this.logToWriter(finalMessage);
  }

  private static logToWriter(text: string): void {
    for (const writer of this.writers) {
      writer(text);
    }
  }
}
